File: CS/NativeXtraReportExample.Module.Win/Controllers/ShowXtraReportController.cs
using System;
using System.Collections.Generic;
using System.Linq;
using DevExpress.ExpressApp;
using NativeXtraReportExample.Module.BusinessObjects;
using DevExpress.ExpressApp.Actions;
using DevExpress.Persistent.Base;
using DevExpress.XtraReports.UI;
using DevExpress.ExpressApp.DC;
using DevExpress.Data.Filtering;
using DevExpress.ExpressApp.Templates;

namespace NativeXtraReportExample.Module.Win.Controllers {
    public class ShowXtraReportController : ObjectViewController<ListView, Task> {
        public ShowXtraReportController() {
            var showReportAction = new SimpleAction(this, "ShowReport", PredefinedCategory.View);
            showReportAction.ImageName = "BO_Report";
            showReportAction.PaintStyle = ActionItemPaintStyle.CaptionAndImage;
            showReportAction.Execute += new SimpleActionExecuteEventHandler(showReportAction_Execute);
            var showReportWithParametersAction = new PopupWindowShowAction(this, "ShowReportWithParameters", PredefinedCategory.View);
            showReportWithParametersAction.ImageName = "BO_Report";
            showReportWithParametersAction.PaintStyle = ActionItemPaintStyle.CaptionAndImage;
            showReportWithParametersAction.CustomizePopupWindowParams += new CustomizePopupWindowParamsEventHandler(showReportWithParameters_CustomizePopupWindowParams);
            showReportWithParametersAction.Execute += new PopupWindowShowActionExecuteEventHandler(showReportWithParameters_Execute);
        }
        void showReportAction_Execute(object sender, SimpleActionExecuteEventArgs e) {
            TaskReport report = new TaskReport();
            report.DataSource = View.CollectionSource.Collection;
            ReportPrintTool tool = new ReportPrintTool(report);
            tool.ShowPreview();
        }
        void showReportWithParameters_CustomizePopupWindowParams(object sender, CustomizePopupWindowParamsEventArgs e) {
            e.DialogController.CanCloseWindow = false;
            e.DialogController.AcceptAction.Caption = "Preview";
            var objectSpace = Application.CreateObjectSpace();
            var parameters = new ReportParameters() { DeadlineBefore = DateTime.Now };
            e.View = Application.CreateDetailView(objectSpace, parameters);
        }
        void showReportWithParameters_Execute(object sender, PopupWindowShowActionExecuteEventArgs e) {
            var parameters = e.PopupWindow.View.CurrentObject as ReportParameters;
            CriteriaOperator criteria = new BinaryOperator("Deadline", parameters.DeadlineBefore, BinaryOperatorType.Less);
            if (!string.IsNullOrEmpty(parameters.SubjectContains)) {
                criteria = GroupOperator.Combine(GroupOperatorType.And, criteria,
                    CriteriaOperator.Parse("Contains([Subject], ?)", parameters.SubjectContains));
            }
            var dataSource = Application.CreateObjectSpace().CreateCollection(typeof(Task), criteria);
            var report = new TaskReport();
            report.DataSource = dataSource;
            report.ShowPreview();
        }
    }
    [DomainComponent]
    public class ReportParameters {
        public string SubjectContains { get; set; }
        public DateTime DeadlineBefore { get; set; }
    }
}

File: VB/NativeXtraReportExample.Module.Win/Controllers/ShowXtraReportController.vb
Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports DevExpress.ExpressApp
Imports NativeXtraReportExample.Module.BusinessObjects
Imports DevExpress.ExpressApp.Actions
Imports DevExpress.Persistent.Base
Imports DevExpress.XtraReports.UI
Imports DevExpress.ExpressApp.DC
Imports DevExpress.Data.Filtering
Imports DevExpress.ExpressApp.Templates

Namespace NativeXtraReportExample.Module.Win.Controllers
    Public Class ShowXtraReportController
        Inherits ObjectViewController(Of ListView, Task)

        Public Sub New()
            Dim showReportAction = New SimpleAction(Me, "ShowReport", PredefinedCategory.View)
            showReportAction.ImageName = "BO_Report"
            showReportAction.PaintStyle = ActionItemPaintStyle.CaptionAndImage
            AddHandler showReportAction.Execute, AddressOf showReportAction_Execute
            Dim showReportWithParametersAction = New PopupWindowShowAction(Me, "ShowReportWithParameters", PredefinedCategory.View)
            showReportWithParametersAction.ImageName = "BO_Report"
            showReportWithParametersAction.PaintStyle = ActionItemPaintStyle.CaptionAndImage
            AddHandler showReportWithParametersAction.CustomizePopupWindowParams, AddressOf showReportWithParameters_CustomizePopupWindowParams
            AddHandler showReportWithParametersAction.Execute, AddressOf showReportWithParameters_Execute
        End Sub
        Private Sub showReportAction_Execute(ByVal sender As Object, ByVal e As SimpleActionExecuteEventArgs)
            Dim report As New TaskReport()
            report.DataSource = View.CollectionSource.Collection
            Dim tool As New ReportPrintTool(report)
            tool.ShowPreview()
        End Sub
        Private Sub showReportWithParameters_CustomizePopupWindowParams(ByVal sender As Object, ByVal e As CustomizePopupWindowParamsEventArgs)
            e.DialogController.CanCloseWindow = False
            e.DialogController.AcceptAction.Caption = "Preview"
            Dim objectSpace = Application.CreateObjectSpace()
            Dim parameters = New ReportParameters() With {.DeadlineBefore = Date.Now}
            e.View = Application.CreateDetailView(objectSpace, parameters)
        End Sub
        Private Sub showReportWithParameters_Execute(ByVal sender As Object, ByVal e As PopupWindowShowActionExecuteEventArgs)
            Dim parameters = TryCast(e.PopupWindow.View.CurrentObject, ReportParameters)
            Dim criteria As CriteriaOperator = New BinaryOperator("Deadline", parameters.DeadlineBefore, BinaryOperatorType.Less)
            If Not String.IsNullOrEmpty(parameters.SubjectContains) Then
                criteria = GroupOperator.Combine(GroupOperatorType.And, criteria, CriteriaOperator.Parse("Contains([Subject], ?)", parameters.SubjectContains))
            End If
            Dim dataSource = Application.CreateObjectSpace().CreateCollection(GetType(Task), criteria)
            Dim report = New TaskReport()
            report.DataSource = dataSource
            report.ShowPreview()
        End Sub
    End Class
    <DomainComponent> _
    Public Class ReportParameters
        Public Property SubjectContains() As String
        Public Property DeadlineBefore() As Date
    End Class
End Namespace